### PR TITLE
Add .with_config() method to Runnables, Add run_id, run_name to RunnableConfig

### PR DESCRIPTION
**Pull request by nfcampos**
- with_config() allows binding any config values to a Runnable, like .bind() does for kwargs

### CODE REVIEW

**Review by hinthornw — COMMENTED** [review]
Very exciting! I guess this means I can do things like this:

```
import uuid

rid = uuid.uuid4()

chain = (
    prompts.ChatPromptTemplate.from_messages([("system", "You are an alien named Reginald Baker"), ("human", "{input}")])
    .with_config({"run_name":"My cool prompt"})
    | chat_models.ChatOpenAI()
)

chain.invoke({"input": "What's the weather like up there?"})
```

I'd imagine at least for run IDs we'd want to pass it in when calling `invoke`, but since we don't prune the run ID when passing to child runs or the next run in a sequence, we end up with duplicates. So for instance this gives errors:

```
import uuid
from langchain import chat_models, prompts

rid = uuid.uuid4()

chain = (
    prompts.ChatPromptTemplate.from_messages([("system", "You are an alien named Reginald Baker"), ("human", "{input}")])
    | chat_models.ChatOpenAI()
)

chain.invoke({"input": "What's the weather like up there?"}, {"run_id": rid})
```

```
{"detail":"Run already exists"}
```

Similarly, run names are propagated

```
import uuid
from langchain import chat_models, prompts, schema

rid = uuid.uuid4()

chain = (
    prompts.ChatPromptTemplate.from_messages([("system", "You are an alien named Reginald Baker"), ("human", "{input}")])
    | chat_models.ChatOpenAI()
    | schema.output_parser.StrOutputParser()
)

chain.invoke({"input": "What's the weather like up there?"}, {"run_name": "My Run"})
```
image

 
Maybe we could filter these out in `patch_config`? Or if it's too confusing which args are inherited and which aren't we could consider something else (maybe making the `with_config()` only local?)

**Comment by nfcampos** [comment]
@hinthornw oof, let me fix that, good catch

**Comment by nfcampos** [comment]
@hinthornw try again now?

File: libs/langchain/langchain/chains/base.py [hunk 1]
```diff
@@ -296,6 +305,8 @@ async def acall(
         *,
         tags: Optional[List[str]] = None,
         metadata: Optional[Dict[str, Any]] = None,
+        run_id: Optional[UUID] = None,
+        run_name: Optional[str] = None,
```
[the remark this answers is not in the document]
Comment (baskaryan) [reply libs/langchain/langchain/chains/base.py:309]: ooc at some pt should we just have this accept the whole config?

**Comment by hinthornw** [comment]
Works now for me! I think LLMs still don't support it fully though.

```
import uuid

run_id = uuid.uuid4()
model = chat_models.ChatOpenAI().with_config({"run_name": "A test model"}).invoke("hi", {"run_id": run_id})

from langsmith import Client

Client().read_run(run_id)
# Raises an error. Run name is also ignored
```

File: libs/langchain/langchain/chains/base.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (nfcampos) [reply libs/langchain/langchain/chains/base.py:303]: Could make sense yea, but this method is exposed as public api too, so I can't easily modify it

File: libs/langchain/langchain/schema/runnable/base.py [hunk 2]
```diff
@@ -210,7 +211,20 @@ def bind(self, **kwargs: Any) -> Runnable[Input, Output]:
...
         Bind arguments to a Runnable, returning a new Runnable.
         """
-        return RunnableBinding(bound=self, kwargs=kwargs)
+        return RunnableBinding(bound=self, kwargs=kwargs, config={})
+
+    def with_config(
+        self,
+        config: Optional[RunnableConfig] = None,
+        # Sadly Unpack is not well supported by mypy so this will have to be untyped
+        **kwargs: Any,
```
Comment (baskaryan) [inline libs/langchain/langchain/schema/runnable/base.py:220]: do we need to support kwargs at all? shouldn't everything just be passed in config, why do we need 2 ways to do same thing

File: libs/langchain/langchain/schema/runnable/base.py [hunk 3]
```diff
@@ -1490,48 +1519,101 @@ def lc_serializable(self) -> bool:
...
-        return self.bound.batch(inputs, config, **{**self.kwargs, **kwargs})
+        configs = cast(
+            List[RunnableConfig],
+            [self._merge_config(conf) for conf in config]
+            if isinstance(config, list)
+            else [
+                patch_config(self._merge_config(config), deep_copy_locals=True)
+                for _ in range(len(inputs))
+            ],
+        )
```
Comment (baskaryan) [inline libs/langchain/langchain/schema/runnable/base.py:1586]: p hard to read, what about
```python
if isinstance(config, list):
    configs = [self._merge_config(conf) for conf in config]
else:
    configs = []
    for _ in range(len(inputs)):
        configs.append(patch_config(self._merge_config(config), deep_copy_locals=True))
```

File: libs/langchain/langchain/schema/runnable/base.py [hunk 4]
```diff
@@ -1490,48 +1519,101 @@ def lc_serializable(self) -> bool:
...
     async def ainvoke(
         self,
         input: Input,
         config: Optional[RunnableConfig] = None,
         **kwargs: Optional[Any],
     ) -> Output:
-        return await self.bound.ainvoke(input, config, **{**self.kwargs, **kwargs})
+        return await self.bound.ainvoke(
+            input,
+            self._merge_config(config),
```
Comment (baskaryan) [inline libs/langchain/langchain/schema/runnable/base.py:1568]: nit should be passed in a named param

File: libs/langchain/langchain/schema/runnable/base.py (hunk 2 above)
Comment (nfcampos) [reply libs/langchain/langchain/schema/runnable/base.py:220]: Because this is easier to use

**Pull request merged**